## OR 506: HomeWork-6
**Created in iPython Notebook by Guduguntla Vamshi <gudugu@ncsu.edu>**

\\begin{equation}
min \ f(x),\ f(x) = x_1^2 + x_2^2 - 6x_1 - 8x_2 + 10
\\end{equation}

\\begin{equation}
constraints \   4x_1^2 + x_2^2 <= 16 \
\ 3x_1 + 5x_2 <= 4 \
\ x_1,x_2 >= 0 \
\\end{equation}

### 1. Exterior Penalty Method:

**Method Used:**

$$
\begin{align*}
    \psi(x,\rho_p) &= J(X) + \rho_p*P(X)\\
    P(X) &= {\sum_{j=1}^{n}max [0,g_j(X)]^2} + {\sum_{k=1}^{m_2} h_k(x)^2}\\
\end{align*}
$$


- if all constraints are satisfied, then **P(x)=0**
- **$ \rho_p $** = penalty parameter; starts as a small number and increases
- if **$ \rho_p $** is small, **$ \psi(x,\rho_p) $** is easy to minimize but yields large constraint violations
- if **$ \rho_p $** is large, constraints are all nearly satisfied but optimization problem is numerically ill-conditioned
- if optimization stops before convergence is reached, the design will be infeasible

**Penalty Function Used:**

$$
\begin{align*}
    \psi(x,\rho_p) &= J(X) + \rho_p*({\sum_{j=1}^{n} [g_j(X)]^2} + {\sum_{k=1}^{m_2} h_k(x)^2})\\
\end{align*}
$$

- **$g_j(X)$** contains those inequality constraints that are violated at x
- It can be shown that as $\rho_p->large$ $lim (x*(\rho_p))=x*$
- $\psi(x,\rho_p)$ is defined everywhere

**Algorithm Used:**

- choose $\rho_0$, set $k=10^0$
- find min $\psi(x,\rho_k)$ => $x_k*$
- if not converged, set $ 10^{k+1} $ > $ 10^{k} $ , k <- k+1 and repeat

**Implementation**

In [14]:
import numpy as np
from sympy import *
import math
import scipy

**Function Definition**

In [2]:
def function():
    x, y=symbols('x y',real=True)
    return x**2 + y**2 - 6*x - 8*y + 10

def constratints():
    x, y=symbols('x y',real=True)
    g1 = 4*x**2 + y**2 - 16
    g2 = 3*x + 5*y - 4
    return g1,g2

def penalized_function(mu,g1,g2):
    return function() + mu * (g1**2 + g2**2)

In [3]:
x,y=symbols('x y',real=True)
g1,g2 = constratints()
soln = np.array(nsolve([diff(penalized_function(1,g1,g2),x),diff(penalized_function(1,g1,g2),y)],[x,y],[0,0]))
function_value = function().subs({x:soln[0],y:soln[1]})

for k in range(20):
    if (g1.subs({x:soln[0],y:soln[1]}) <= 0) and (g2.subs({x:soln[0],y:soln[1]}) > 0):
        F = penalized_function(10**k,0,g2)
        soln = np.array(nsolve([diff(penalized_function(10**k,0,g2),x),diff(penalized_function(10**k,0,g2),y)],[x,y],[0,0]))
        function_value = function().subs({x:soln[0],y:soln[1]})
        print 'iteration',k,'Mu: ',10**k,'x1 x2 = ',round(soln[0],9),round(soln[1],9),'f(x1,x2): ',round(function_value,9)
    elif (g1.subs({x:soln[0],y:soln[1]}) > 0) and (g2.subs({x:soln[0],y:soln[1]}) <= 0):
        F = penalized_function(10**k,g1,0)
        soln = np.array(nsolve([diff(penalized_function(10**k,g1,0),x),diff(penalized_function(10**k,g1,0),y)],[x,y],[0,0]))
        function_value = function().subs({x:soln[0],y:soln[1]})
        print 'iteration',k,'Mu: ',10**k,'x1 x2 = ',round(soln[0],9),round(soln[1],9),'f(x1,x2): ',round(function_value,9)
    else:
        print 'Constraints Satisfied at iteration: ',k
        print 'Final Solution is :'
        print 'Mu: ',10**k,'x1 x2 = ',round(soln[0],9),round(soln[1],9),'f(x1,x2): ',round(function_value,9)
        break

iteration 0 Mu:  1 x1 x2 =  0.857142857 0.428571429 f(x1,x2):  2.346938776
iteration 1 Mu:  10 x1 x2 =  0.80058651 0.33431085 f(x1,x2):  3.27469664
iteration 2 Mu:  100 x1 x2 =  0.794766245 0.324610409 f(x1,x2):  3.371544561
iteration 3 Mu:  1000 x1 x2 =  0.794182524 0.32363754 f(x1,x2):  3.381271674
iteration 4 Mu:  10000 x1 x2 =  0.794124135 0.323540225 f(x1,x2):  3.38224481
iteration 5 Mu:  100000 x1 x2 =  0.794118296 0.323530493 f(x1,x2):  3.382342128
iteration 6 Mu:  1000000 x1 x2 =  0.794117712 0.32352952 f(x1,x2):  3.38235186
iteration 7 Mu:  10000000 x1 x2 =  0.794117654 0.323529423 f(x1,x2):  3.382352833
iteration 8 Mu:  100000000 x1 x2 =  0.794117648 0.323529413 f(x1,x2):  3.38235293
iteration 9 Mu:  1000000000 x1 x2 =  0.794117647 0.323529412 f(x1,x2):  3.38235294
iteration 10 Mu:  10000000000 x1 x2 =  0.794117647 0.323529412 f(x1,x2):  3.382352941
iteration 11 Mu:  100000000000 x1 x2 =  0.794117647 0.323529412 f(x1,x2):  3.382352941
iteration 12 Mu:  1000000000000 x1 x2 =  

**Comments on Results:**

- Convergence criteria is: 20 maximum iterations or **constraints satisfied**
- Using the initial value as Mu = 10^0 = 1 final solution of x is **[0.794117655,0.323529407]**
- The solution is obtained at 2 iterations at **Mu = 100**, x is **[0.794766245,0.324610409]**
- However, allowing for tolerance in the constraints around **10^-8**, the iterations are further increased
- Mu values in **16 iterations** increase to the powers of 10.
- The converging function value is **3.382352941**
- The algorithm converged in **16 iterations** using the above mu values and tolerance of **10^-8** in constraints
- The computed values are cross-validated by checking plugging in these values given by the criteria defined above and   checking for constraints.

### 2. Interior Penalty Method:

**Method Used:**

$$
\begin{align*}
    \psi(x,\rho_p,r_p) &= J(X) + \rho_p*P(X)\\
    P(X) &= r_p * {\sum_{j=1}^{n} - ln [g_j(X)]} + \rho_p*{\sum_{k=1}^{m_2} h_k(x)^2}\\
\end{align*}
$$

- $r_p$ = barrier parameter; starts as a large positive number and decreases
- barrier function is defined for inequality constraints only
- sequence of improving feasible designs
- $psi(x,\rho_p,r_p)$ discontinuous at constraint boundaries



**Penalty Function Used:**

$$
\begin{align*}
    P(X) &= {\sum_{j=1}^{m_!} - ln [-g_j(X)]}\\
    \psi(x,\rho_p,r_p) &= J(X) - \rho_p*{\sum_{j=1}^{m_!} - ln [-g_j(X)]}\\
\end{align*}
$$



- penalty function has a positive singularity at the boundary of the feasible region
- penalty function is undefined for gi(x)>0
- It can be shown that as $r_p->0$ $lim (x*(r_p))=x*$

**Algorithm Used:**

- choose $\rho_0$, set $k=10^0$
- find min $\psi(x,\rho_k)$ => $x_k*$
- if not converged, set $ 10^{-(k+1)} $ < $ 10^{k} $ , k <- k+1 and repeat

**Implementation**

In [19]:
import numpy as np
from sympy import *
import math
import scipy

In [20]:
def function():
    x, y=symbols('x y',real=True)
    return x**2 + y**2 - 6*x - 8*y + 10

def constratints():
    x, y=symbols('x y',real=True)
    g1 = (-1)*(4*x**2 + y**2 - 16)
    g2 = (-1)*(3*x + 5*y - 4)
    return g1,g2

def penalized_function(mu,g1,g2):
    return function() - mu * (log(g1) + log(g2))

In [21]:
x,y=symbols('x y',real=True)
g1,g2 = constratints()
soln = np.array(nsolve([diff(penalized_function(1,g1,g2),x),diff(penalized_function(1,g1,g2),y)],[x,y],[.6,.3]))
function_value = function().subs({x:soln[0],y:soln[1]})
print soln,function_value
k = 0

soln0 = np.array([0,0])

while k < 85:
    k += 1
    F = penalized_function(1-.01*k,g1,g2)
    soln = np.array(nsolve([diff(penalized_function(1-.01*k,g1,g2),x),diff(penalized_function(1-.01*k,g1,g2),y)],[x,y],[.6,.3]))
    function_value = function().subs({x:soln[0],y:soln[1]})
    print 'iteration',k,'Mu: ',1-.01*k,'x1 x2 = ',round(soln[0],9),round(soln[1],9),'f(x1,x2): ',round(function_value,9)
    
    if np.linalg.norm(soln-soln0) < 0.001:
        break
    else:
        soln0 = soln
        continue

[mpf('0.61714266168678841') mpf('0.29404825032408209')] 4.41208746567912
iteration 1 Mu:  0.99 x1 x2 =  0.618573055 0.294534793 f(x1,x2):  4.401666694
iteration 2 Mu:  0.98 x1 x2 =  0.620008585 0.295018625 f(x1,x2):  4.391246127
iteration 3 Mu:  0.97 x1 x2 =  0.621449288 0.29549972 f(x1,x2):  4.380825813
iteration 4 Mu:  0.96 x1 x2 =  0.622895201 0.295978054 f(x1,x2):  4.370405801
iteration 5 Mu:  0.95 x1 x2 =  0.624346362 0.296453601 f(x1,x2):  4.359986138
iteration 6 Mu:  0.94 x1 x2 =  0.62580281 0.296926334 f(x1,x2):  4.349566876
iteration 7 Mu:  0.93 x1 x2 =  0.627264582 0.297396227 f(x1,x2):  4.339148065
iteration 8 Mu:  0.92 x1 x2 =  0.628731718 0.297863253 f(x1,x2):  4.328729754
iteration 9 Mu:  0.91 x1 x2 =  0.630204258 0.298327387 f(x1,x2):  4.318311997
iteration 10 Mu:  0.9 x1 x2 =  0.63168224 0.298788599 f(x1,x2):  4.307894844
iteration 11 Mu:  0.89 x1 x2 =  0.633165707 0.299246863 f(x1,x2):  4.29747835
iteration 12 Mu:  0.88 x1 x2 =  0.634654698 0.299702151 f(x1,x2):  4.287

**Comments on Results:**

- Convergence criteria is: 100 maximum iterations or **constraints satisfied**
- Using the initial value as Mu = 10^0 = 1 final solution of x is **[0.761919857,0.322434382]**
- The solution is obtained at 85 iterations at **Mu = 0.15**, x is **[0.761919857,0.322434382]**
- However, allowing for tolerance in the constraints around **10^-8**, the iterations are further increased
- Mu values in **16 iterations** decrease by *1-.01*k* as k is the iteration number.
- The converging function value is **3.533491602**
- The algorithm converged in **85 iterations** using the above mu values and tolerance of **10^-8** in constraints
- The computed values are cross-validated by checking plugging in these values given by the criteria defined above and   checking for constraints.

### 3. Zoutendijk Method:

**Implementation**

In [ ]:
%%python%%

def function():
    x, y=symbols('x y',real=True)
    return x1**2 + x2**2 - 6*x1 - 8*x2 + 10

def constratints():
    x, y=symbols('x1 x2',real=True)
    g1 = 4*x1**2 + x2**2 - 16
    g2 = 3*x1 + 5*x2 - 4
    g3 = -1*x1
    g4 = -1*x2
    return g1,g2,g3,g4

f = function()
g1,g2,g3,g4 = constratints()

In [ ]:
%%matlab%%

syms x1 x2
x = [0 0]
g = [g1;g2;g3;g4]
eps = 10^-8
xfinal = Zoutendijks_Method(f,g,x,eps,20)

In [ ]:
# define the function
function xopt = Zoutendijks_Method(f , g , x_1, eps,maxit)
    itr = 0
    sze = size(g,1);
varlist = symvar(f);
    while(itr < maxit)
        c=0;
        itr = itr + 1;
        for i = 1:sze
            if(subs(g(i,1) , varlist , x_1)<0)
                c=c+1;
            end
        end

        if(c==sze)
            S = double(-1*subs(gradient(f,varlist),varlist,x_1));
            S = S/norm(S);
            lam = zout_optlam(f,g,S,x_1);
            x = x_1 + lam*S.';
            if (abs((subs(f, varlist, x_1)-subs(f, varlist, x))/subs(f, varlist, x_1))<=eps && norm(x_1-x)<=eps)
                xopt = x;
                break;
            end
            x_1 = x
            continue;
        end

        fcons = double([subs(gradient(f, varlist),varlist,x_1)]).';

        acoe = [];
        for i = 1:sze
            if(subs(g(i,1) , varlist , x_1)==0)
                ccoe = [subs(gradient(g(i,1), varlist),varlist,x_1)].';
                acoe = [acoe;ccoe];
            end
        end
        A = [acoe;fcons];
        A = double(A);
        A = [ones(size(A,1),1) A];
        B = zeros(size(A,1),1);
        fl = [-1 zeros(1,size(A,2)-1)];
        lb = [-Inf -1*ones(1,size(A,2)-1)];
        ub = [Inf ones(1,size(A,2)-1)];
        xlp = linprog(fl , A , B , [] , [] , lb , ub);

        if(abs(xlp(1,1))<eps)
            xopt = x_1;
            break;
        end

        S = double(xlp(2:size(xlp,1),1));
        S = S/norm(S);
        lam = zout_optlam(f,g,S,x_1);
        x = x_1 + lam*S.';
        if (abs((subs(f, varlist, x_1)-subs(f, varlist, x))/subs(f, varlist, x_1))<=eps && norm(x_1-x)<=eps)
            xopt = x;
            break;
            end
        x_1 = x
    end
end

In [ ]:
"""
iteration  x1       x2       f(x1,x2)
1	0.353553391	0.353553391	5.300252527
2	0.426974751	0.454718126	4.189482498
3	0.463685432	0.505300494	3.645816225
4	0.468274267	0.51162329	3.578407258
5	0.470568684	0.514784688	3.544748553
6	0.471715893	0.516365387	3.527930643
7	0.472002695	0.516760562	3.523727357
8	0.47202062	0.516785261	3.523464664
9	0.472021741	0.516786804	3.523448247
10	0.472021811	0.516786901	3.523447217
60	0.793624047	0.323825246	3.382355668
61	0.793624169	0.323825451	3.382353622
62	0.793833508	0.323699825	3.382353567
63	0.794042851	0.323574206	3.382353562
64	0.794095188	0.323542803	3.382353561
65	0.794108272	0.323534952	3.382353565
66	0.794114814	0.323531027	3.382353563
67	0.79411645	0.323530046	3.382353559
68	0.794117268	0.323529555	3.38235356
"""

**Comments on Results:**

- final solution of x is **[0.794117268,0.323529555]**
- However, allowing for tolerance in the constraints around **10^-8**, the iterations are further increased
- The converging function value is **3.38235356**
- The algorithm converged in **68 iterations** 
- The computed values are cross-validated by checking plugging in these values given by the criteria defined above and  checking for constraints.